src/adaptivemvnormal.jl

```diff
@@ -0,0 +1,84 @@
+"""
+    AdaptiveMvNormal(constant_component::MvNormal; σ=2.38, β=0.05)
+
+Adaptive multivariate normal mixture proposal as described in Haario et al. and
+Roberts & Rosenthal (2009). Uses a two-component mixture of MvNormal
+distributions. One of the components (with mixture weight `β`) remains
+constant, while the other component is adapted to the target covariance
+structure. The proposal is initialized by providing the constant component to
+the constructor. 
+
+`σ` is the scale factor for the covariance matrix, where 2.38 is supposedly
+optimal in a high-dimensional context according to Roberts & Rosenthal.
+
+# References
+
+- Haario, Heikki, Eero Saksman, and Johanna Tamminen. 
+  "An adaptive Metropolis algorithm." Bernoulli 7.2 (2001): 223-242.
+- Roberts, Gareth O., and Jeffrey S. Rosenthal. "Examples of adaptive MCMC."
+  Journal of Computational and Graphical Statistics 18.2 (2009): 349-367.
+"""
+mutable struct AdaptiveMvNormal{T1,T2,V} <: Proposal{T1}
+    d::Int  # dimensionality
+    n::Int  # iteration
+    β::Float64  # constant component mixture weight
+    σ::Float64  # scale factor for adapted covariance matrix
+    constant::T1  
+    adaptive::T2
+    Ex::Vector{V}  # rolling mean vector
+    EX::Matrix{V}  # scatter matrix of previous draws
+end
+
+function AdaptiveMvNormal(dist::MvNormal; σ=2.38, β=0.05)
+    n = length(dist)
+    adaptive = MvNormal(PDMat(Symmetric(dist.Σ)))
+    AdaptiveMvNormal(n, -1, β, σ, dist, adaptive, zeros(n), zeros(n,n))
+end
+
+is_symmetric_proposal(::AdaptiveMvNormal) = true
+
+"""
+    adapt!(p::AdaptiveMvNormal, x::AbstractVector)
+
+Adaptation for the adaptive multivariate normal mixture proposal as described
+in Haario et al. (2001) and Roberts & Rosenthal (2009). Will perform an online
+estimation of the target covariance matrix and mean. The code for this routine
+is largely based on `Mamba.jl`.
+"""
+function adapt!(p::AdaptiveMvNormal, x::AbstractVector)
+    p.n += 1
+    # adapt mean vector and scatter matrix
+    f = p.n / (p.n + 1.0)
+    p.Ex = f * p.Ex + (1.0 - f) * x
+    p.EX = f * p.EX + (1.0 - f) * x * x'
+    # compute adapted covariance matrix
+    Σ = (p.σ^2 / (p.d * f)) * (p.EX - p.Ex * p.Ex') 
+    F = cholesky(Hermitian(Σ), check=false) 
+    if rank(F.L) == p.d
+        p.adaptive = MvNormal(PDMat(Σ, F))
+    end
```

Comment:
Also here, can't we just use the default constructors of `MvNormal` which call `PDMat` implicitly?
```suggestion
    p.adaptive = MvNormal(Σ)
```

Reply:
I implemented it this way (inspired by the Mamba implementation) to check whether the empirical estimate of the covariance matrix is a proper covariance matrix. If it is we can adapt the proposal distribution without additional computation by feeding the cholesky decomposition to MvNormal this way. How would you do it otherwise, I guess you could use `try/catch` but that doesn't seem nicer?

Reply:
No, I don't think a `try`-`catch` block would be any better. IMO it is not good to just suppress the errors here and use an incorrect Cholesky decomposition with possibly negative eigenvalues. I would suggest the following:
- If possible, update the Cholesky decomposition (or some other parametrization of the covariance matrix) iteratively. Usually this is more efficient than recomputing the factorization in every step and it avoid the numerical problems.
- Often calling `Matrix(Symmetric(Sigma))` is sufficient to avoid problems with matrices that are not completely symmetric and for which therefore the Cholesky decomposition fails.
- Alternatively, use [`PositiveFactorizations`](https://github.com/timholy/PositiveFactorizations.jl).

IMO the first alternative would clearly be the best, if possible.

Reply:
I just had a quick look at Haario et al. and the update formula of the covariance matrix in eq (3). It seems that indeed the update could be performed by only `d + 3` rank 1 up- and downdates which each can be performed in O(d^2). Actually, if `eps = 0`, then only 3 rank 1 up- and downdates are needed and therefore the iterative update of the Cholesky decomposition would be much cheaper than recomputing the decomposition in O(d^3) operations. The rank 1 up- and downdates are implemented in `LinearAlgebra.lowrankupdate!` and `LinearAlgebra.lowrankdowndate!` (and their corresponding out-of-place methods).

I think this would be the preferable method for updating the Cholesky decomposition. Probably it would also be cheaper to save just the Cholesky decomposition and not wrap it in a `MvNormal` object since it is trivial to sample from the multivariate normal distribution with the Cholesky decomposition.

Reply:
That sounds interesting and worthwhile, have to look at it. FWIW I just based my code on the description in the Roberts & Rosenthal paper and the Mamba implementation, since I simply wanted to use this for a problem I'm having.

I had the impression from the code that there is a preference to use Distributions.jl distributions for sampling, that's why I wrapped them in MvNormal objects instead of using the Cholesky factor directly (when both the full matrix and cholesky factors are available, it is as cheap to put them in an MvNormal).